packages/toolpad-app/src/runtime/ToolpadApp.tsx

```diff
@@ -268,6 +270,30 @@ function RenderedNodeContent({ node, childNodeGroups, Component }: RenderedNodeC
     return hookResult;
   }, [isLayoutNode, liveBindings, nodeId]);
 
+  const changeClosestFormValue = React.useCallback(
```

Comment:
Not sure what the specific issue is for the missing default value - you'd probably have to look into the bindings logic in the runtime to figure it out - there's a `useEffect` there that handles default values. (`src/runtime/ToolpadApp`, around line 349?). At least this is where i would look first.

There probably are some similarities between what you need to do here and what I did for the List component in https://github.com/mui/mui-toolpad/pull/1527 (looking for a review there btw to see if my latest method sounds good? Also maybe it would be helpful if we merged the List first.).
What I did there was to create a local scope inside the list component by making it set a context, on the runtime (`src/runtime/ToolpadApp`) that wraps all of it's children. In the case of a form, this context could probably hold the form state.

Reply:
How I would go about it without doing low-level work to toolpad to enable defining arbitrary scopes. 

1. create shared `FormValue` and `SetFormValue` contexts to be used by both the Form and individual components. These contexts are provided by the `Form` component to provide its `value` and `onChange` prop downstream:

```tsx
<Paper>
  <FormValue.Provider value={value}>
    <SetFormValue.Provider
      value={(field, newValue) => onChange({ ...value, [field]: newValue })}
    >
      {children}
    </SetFormValue.Provider>
  </FormValue.Provider>
</Paper>
```

In the form components we use these contexts, and if they exist we override the `value` and `onChange` from the props:

```tsx
const formValue = React.useContext(FormValue);
const setFormValue = React.useContext(SetFormValue);
inveriant(!!formValue === !!setFormValue, "...");

const inputValue = formValue ? formValue[name] : valueProp;
const setInput = setFormValue
  ? (newValue) => setFormValue(name, newValue)
  : onChangeProp;

return <TextField value={input} onChange={setInput} />;
```

2. If the above works, we have the basis, but we still need to filter out the components that are bound to a form in the bindings. The second part of 1. will be shared by every form-capable component, so we can extract it and put it in the runtime. e.g. we can add some definition to `ToolpadComponentConfig` like:

```ts
argTypes: {
  /* ... */
},
formValue: ['value', 'onChange']
```

In the runtime we can check for existence of this config when rendering the component and do the logic of 1. automatically, when preparing its properties.
Then we can also read it while setting up the bindings and omit these properties from creating scope variables (so just remove `scopePath` for those)


This is a two part explanation, the end result is two, but do 1. first to verify that the reasoning works.

Reply:
@Janpot 's idea sounds good to me!

> Could you please clarify what you mean by that? Currently it's implemented in a way that child components wouldn't have to know anything about form component

In this case this `onChange` handler would probably need to be passed as a prop to each component.

I guess one of the main advantages of using context would be that you wouldn't have to "search" for the closest form manually.
As for including form logic inside components themselves, I'm fine with that if we decide it's the easiest/best way to go about it.